File: src/main/java/com/sleepkqq/sololeveling/config/GrpcConfigClient.java
package com.sleepkqq.sololeveling.config;

import com.sleepkqq.sololeveling.proto.config.DefaultGrpcClientConfig;
import com.sleepkqq.sololeveling.proto.task.TaskServiceGrpc;
import com.sleepkqq.sololeveling.proto.task.TaskServiceGrpc.TaskServiceBlockingStub;
import com.sleepkqq.sololeveling.proto.user.UserServiceGrpc;
import com.sleepkqq.sololeveling.proto.user.UserServiceGrpc.UserServiceBlockingStub;
import io.grpc.ManagedChannel;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class GrpcConfigClient extends DefaultGrpcClientConfig {

  public GrpcConfigClient(@Value("${app.host}") String host) {
    super(host);
  }

  @Bean
  public ManagedChannel userServiceChannel(@Value("${app.grpc.services.user.port}") int port) {
    return createManagedChannel(port);
  }

  @Bean
  public UserServiceBlockingStub userServiceBlockingStub(
      ManagedChannel userServiceChannel
  ) {
    return UserServiceGrpc.newBlockingStub(userServiceChannel);
  }

  @Bean
  public ManagedChannel taskServiceChannel(@Value("${app.grpc.services.task.port}") int port) {
    return createManagedChannel(port);
  }

  @Bean
  public TaskServiceBlockingStub taskServiceBlockingStub(
      ManagedChannel taskServiceChannel
  ) {
    return TaskServiceGrpc.newBlockingStub(taskServiceChannel);
  }
}

File: solo-leveling-gateway-service/src/main/kotlin/com.sleepkqq.sololeveling/gateway/config/GrpcConfigClient.kt
package com.sleepkqq.sololeveling.gateway.config

import com.sleepkqq.sololeveling.gateway.config.properties.GrpcPlayerServiceProperties
import com.sleepkqq.sololeveling.proto.config.DefaultGrpcClientConfig
import com.sleepkqq.sololeveling.proto.player.PlayerServiceGrpc
import com.sleepkqq.sololeveling.proto.player.PlayerServiceGrpc.PlayerServiceBlockingStub
import com.sleepkqq.sololeveling.proto.user.UserServiceGrpc
import com.sleepkqq.sololeveling.proto.user.UserServiceGrpc.UserServiceBlockingStub
import io.grpc.ManagedChannel
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Suppress("unused")
@Configuration
@EnableConfigurationProperties(GrpcPlayerServiceProperties::class)
class GrpcConfigClient(
	private val properties: GrpcPlayerServiceProperties
) : DefaultGrpcClientConfig() {

	@Bean
	fun playerManagedChannel(): ManagedChannel = createManagedChannel(properties)

	@Bean
	fun userServiceBlockingStub(
		playerManagedChannel: ManagedChannel
	): UserServiceBlockingStub = UserServiceGrpc.newBlockingStub(playerManagedChannel)

	@Bean
	fun playerServiceBlockingStub(
		playerManagedChannel: ManagedChannel
	): PlayerServiceBlockingStub = PlayerServiceGrpc.newBlockingStub(playerManagedChannel)
}
